presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java

```diff
@@ -0,0 +1,221 @@
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private static final String INTERSECT_MARKER = "intersect_marker";
+        private static final Signature COUNT_AGGREGATION = new Signature("count", AGGREGATE, "bigint", "boolean");
+        private final PlanNodeIdAllocator idAllocator;
+        private final SymbolAllocator symbolAllocator;
+
+        private Rewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitIntersect(IntersectNode node, RewriteContext<Void> rewriteContext)
+        {
+            List<PlanNode> sources = node.getSources().stream()
+                    .map(rewriteContext::rewrite)
+                    .collect(toList());
+
+            List<Symbol> markers = allocateMarkers(sources.size());
+
+            // identity projection for all the fields in each of the sources plus marker columns
+            List<PlanNode> withMarkers = appendMarkers(markers, sources, node);
+
+            // add a union over all the rewritten sources. The outputs of the union have the same name as the
+            // original intersect node
+            List<Symbol> outputs = node.getOutputSymbols();
+            UnionNode union = union(withMarkers, ImmutableList.copyOf(concat(outputs, markers)));
+
+            // add count aggregations and filter rows where any of the counts is >= 1
+            AggregationNode aggregation = computeCounts(union, outputs, markers);
+            FilterNode filterNode = addFilter(aggregation);
+
+            return project(filterNode, outputs);
+        }
+
+        private List<Symbol> allocateMarkers(int count)
+        {
+            ImmutableList.Builder<Symbol> markers = ImmutableList.builder();
+            for (int i = 0; i < count; i++) {
+                markers.add(symbolAllocator.newSymbol(INTERSECT_MARKER, BOOLEAN));
+            }
+            return markers.build();
+        }
+
+        private List<PlanNode> appendMarkers(List<Symbol> markers, List<PlanNode> nodes, IntersectNode intersect)
+        {
+            ImmutableList.Builder<PlanNode> result = ImmutableList.builder();
+            for (int i = 0; i < nodes.size(); i++) {
+                result.add(appendMarkers(nodes.get(i), i, markers, intersect.sourceSymbolMap(i)));
+            }
+            return result.build();
+        }
+
+        private PlanNode appendMarkers(PlanNode source, int markerIndex, List<Symbol> markers, Map<Symbol, QualifiedNameReference> projections)
+        {
+            ImmutableMap.Builder<Symbol, Expression> assignments = ImmutableMap.builder();
+            // add existing intersect symbols to projection
+            for (Symbol intersectSymbol : projections.keySet()) {
```

Comment:
Thanks @martint for the suggestions in reusing the symbols. I updated the class with your comments. Couple of updates made here though:

1. Create new symbols for the existing projections. Since the intersectSymbol is not unique between the sources and hence the `UnaliasSymbolReferences` rewriter gets confused when there are duplicates in assignments.values(). 

2. Also allocated new symbol for markers in assignments (line 163). Otherwise `UnaliasSymbolReferences` rule for UnionNode `canonicalize` all the reused markers into  one mapping (thus creating n ^ 2 entries in `outputToInputs` instead of n, when there are `n` sources)